### DISCUSSION TITLE
Better support for non-default-constructable datatypes

### DISCUSSION THREAD

**Discussion by Ericson2314** [nlohmann/json#2335] [Ideas]
I have read the docs on how this library does in fact support deserializing to non-default-constructable types. The problem is that it's not very compositional: for example, there is no way to write a generic instance for `std::variant<Ts..>` that  works if all of `Ts..` aren't default construable. Even `std::option<T>`, which is itself default construable no matter if `T` is, would be a bit hard to write in an exception-safe way.

In my own code (https://github.com/NixOS/nix/pull/3859/files#diff-48ed85ec1c3c2412447bedb0e859275a if anyone is curious) I tried a different technique to get around this, which adopted to this library would look something like
```c++
/* To guide overloading */
template<typename T>
struct Proxy {};

V from_json(const json& j, Proxy<V> _);
```
the idea is that `Proxy` guides overloading so we can effectively overload the return type.

**Comment by nlohmann** [comment]
Maybe related: https://github.com/nlohmann/json/pull/2576

**Comment by Ericson2314** [comment]
The ADL `from_json` indeed works quite nicely, and we've been using it!
